### PR TITLE
Uniform subtype header

### LINKED ISSUE #78: Data draft: E2E bit field and data ordering
Order of data appearing should follow the bit ordering in the type.
Similar to text for trace type:
" The order of packing the data fields in each node data element follows the bit order of the IOAM-Trace-Type field"

### PR DESCRIPTION
**Pull request by inband-oam**
Defines the subtype field in a 4 octet header for POT and E2E
Fixes #78
Expands E2E subtype to 16bits

### CODE REVIEW

**Comment by brockners** [comment]
Looks good to me.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 1]
```diff
@@ -1070,15 +1070,64 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+   |IOAM POT Type  | IOAM POT flags|           Reserved            |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+   IOAM proof of transit option data MUST be 4-octet aligned.:
 
-   IOAM proof of transit option data MUST be 4-octet aligned:
+     0                   1                   2                   3
+     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+    |       POT Option data field determined by IOAM-POT-Type       ~
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: I assume '~' at the end is a typo?
Also wondering why this is indented one more space than a few lines above.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 2]
```diff
@@ -1070,15 +1070,64 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+
+
+
+   IOAM POT Type:  8-bit identifier of a particular POT variant that
+      dictates the POT data that is included.  This document defines POT
+      Type 0:
+
+      0: POT data is a 16 Octet field as described below.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: It is a little awkward that this does not have a name. Not that I have any suggestions for that name ...

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 3]
```diff
@@ -1070,15 +1070,64 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+
+   IOAM POT flags:  8-bit.  Following flags are defined:
+
+      Bit 0  "Profile-to-use" (P-bit) (most significant bit).  Indicates
+         which POT-profile is used by POT Type 0.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: I don't think this needs to be specific to POT Type 0 only.

Bit 0  "Profile-to-use" (P-bit) (most significant bit).
For IOAM POT types that use a maximum of two profiles to drive computation, indicates which POT-profile is used. The two profiles are numbered 0, 1.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 4]
```diff
@@ -1070,15 +1070,64 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+
+      Bit 1-7  Reserved: Must be zero.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: Must be set to zero upon transmission and ignored upon receipt.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 5]
```diff
@@ -1070,15 +1070,64 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+
+   Reserved:  16-bit Reserved bits are present for future use.  The
+      reserved bits MUST be set to 0x0.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: Must be set to zero upon transmission and ignored upon receipt.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 6]
```diff
@@ -1091,17 +1140,22 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
 
 
 
-   IOAM POT Type:  7-bit identifier of a particular POT variant that
-      dictates the POT data that is included.  This document defines POT
-      Type 0:
+   IOAM POT Type:  8-bit identifier of a particular POT variant that
+      dictates the POT data that is included, to be set to 0 for the POT
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: The wording is a little difficult to follow.

8-bit identifier of a particular POT variant that dictates the POT data that is included.  This section defines the POT data when the IOAM POT Type is set to the value 0.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 7]
```diff
@@ -1091,17 +1140,22 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+   R (7 bits):  7-bit IOAM POT flags for future use.  MUST be zero on
+      transmission and ignored on receipt.
 
-   Profile to use (P):  1-bit.  Indicates which POT-profile is used to
-      generate the Cumulative.  Any node participating in POT will have
-      a maximum of 2 profiles configured that drive the computation of
-      cumulative.  The two profiles are numbered 0, 1.  This bit conveys
-      whether profile 0 or profile 1 is used to compute the Cumulative.
+   Reserved:  16-bit Reserved bits are present for future use.  The
+      reserved bits MUST be set to 0x0.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: set to 0 upon transmission and ignored upon receipt.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 8]
```diff
@@ -1138,10 +1194,11 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
 
       IOAM edge-to-edge option header:
 
-       0 1 2 3 4 5 6 7
-      +-+-+-+-+-+-+-+-+
-      | IOAM-E2E-Type |
-      +-+-+-+-+-+-+-+-+
+      0                   1                   2                   3
+      0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: The numbers need to be indented by one space

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 9]
```diff
@@ -1152,8 +1209,9 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
       +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 
 
-   IOAM-E2E-Type:  8-bit identifier of a particular in situ OAM E2E
-      variant.
+   IOAM-E2E-Type:  16-bit identifier of a particular in situ OAM E2E
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: A 16-bit identifier which specifies which data types are used in the E2E option data. The IOAM-E2E-Type value is a bit field. The order of packing ...

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 10]
```diff
@@ -1200,6 +1258,15 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
                If the NTP format is used the analyzer should correlate
                several packets in order to detect the error indication.
 
+      Bit 4-15 Undefined.  An IOAM encapsulating node must set the value
+               of each of these bits to 0.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: upon transmission and ignore upon receipt.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 11]
```diff
@@ -1200,6 +1258,15 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
...
+
+   Reserved:  16-bits Reserved bits are present for future use.  The
+      reserved bits MUST be set to 0x0.
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: upon transmission and ignored upon receipt.

File: drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt [hunk 12]
```diff
@@ -1472,7 +1544,33 @@ Internet-Draft           In-situ OAM Data Fields           February 2018
 
    The subsequent sub-sections detail the registries herein contained.
 
-7.2.  IOAM Trace Type Registry
+7.2.  IOAM Type Registry
+
+   This registry defines 128 code points to define IOAM-Type field for
```
Comment (mickeyspiegel) [inline drafts/versions/02/draft-ietf-ippm-ioam-data-02.txt]: This registry defines 128 code points for the IOAM-Type field for ...

**Comment by inband-oam** [comment]
Edited to address the comments raised by Mickey.
Except the transport draft reference. PErhaps we should not have a reference to any of the encap drafts in the data draft at all to avoid later challenges while progressing the draft.

**Pull request merged**